pkg/graph/constructor.go

```diff
@@ -457,18 +505,31 @@ func getSources(ctx context.Context, config ConstructorConfig, logger zap.Logger
 
 		for _, ns := range config.Namespaces {
 			sourcesList, err := config.DynamicClient.Resource(sourceGVR).Namespace(ns).List(ctx, metav1.ListOptions{})
-			if err != nil {
-				// just log and continue, we may succeed for other sources
-				logger.Warn("Failed to list sources", zap.Error(err))
+
+			if apierrs.IsNotFound(err) {
+				continue
+			}
+
+			if apierrs.IsUnauthorized(err) || apierrs.IsForbidden(err) {
+				if !config.Lenient {
+					return nil, fmt.Errorf("failed to list sources: %w", err)
+				}
+				logger.Warn("failed to list sources while constructing lineage graph", zap.Error(err))
 				continue
 			}
 
+			if err != nil {
+				return nil, fmt.Errorf("failed to list sources: %w", err)
+			}
+
 			for i := range sourcesList.Items {
 				unstructuredSource := sourcesList.Items[i]
 				duckSource, err := duckSourceFromUnstructured(&unstructuredSource)
-				if err == nil {
-					duckSources = append(duckSources, duckSource)
+				if err != nil {
+					return nil, fmt.Errorf("failed to convert unstructured source to duck source: %w", err)
 				}
+
+				duckSources = append(duckSources, duckSource)
 			}
 		}
```

Comment:
I made a behavior change here.

Before:
- Ignore the issue that the unstructrued source cannot be converted to a duck source

After:
- Do not ignore it, even with `Lenient=true`